analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs

```diff
@@ -102,135 +88,60 @@ protected override void Initialize(SonarAnalysisContext context)
-                    return syntax?.Identifier.GetLocation();
-                },
-                "method");
-        }
-
-        private static void CheckEvent(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IEventSymbol @event)
-        {
-            CheckEventOrMethod(@event, containerClassSymbol, context,
-                e =>
-                {
-                    var reference = e.DeclaringSyntaxReferences.FirstOrDefault();
-                    if (reference == null)
-                    {
-                        return null;
-                    }
-
-                    if (reference.GetSyntax() is VariableDeclaratorSyntax variableSyntax)
-                    {
-                        return variableSyntax.Identifier.GetLocation();
-                    }
-
-                    var eventSyntax = reference.GetSyntax() as EventDeclarationSyntax;
-                    return eventSyntax?.Identifier.GetLocation();
-                },
-                "event");
-        }
-
-        private static void CheckField(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IFieldSymbol field)
+        private static void CheckFieldOrProperty<T>(T propertyOrField,
+                                                    INamedTypeSymbol containerClassSymbol,
+                                                    SymbolAnalysisContext context,
+                                                    string memberType) where T : ISymbol
         {
-            CheckFieldOrProperty(field, containerClassSymbol, context,
-                f =>
-                {
-                    var reference = f.DeclaringSyntaxReferences.FirstOrDefault();
-                    var syntax = reference?.GetSyntax() as VariableDeclaratorSyntax;
-                    return syntax?.Identifier.GetLocation();
-                },
-                "field");
-        }
-
-        private static void CheckProperty(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IPropertySymbol property)
-        {
-            CheckFieldOrProperty(property, containerClassSymbol, context, p =>
-            {
-                var reference = p.DeclaringSyntaxReferences.FirstOrDefault();
-                if (reference == null)
-                {
-                    return null;
-                }
-                if (!(reference.GetSyntax() is PropertyDeclarationSyntax syntax))
-                {
-                    return null;
-                }
-                return syntax.Identifier.GetLocation();
-            }, "property");
-        }
-
-        private static void CheckFieldOrProperty<T>(T propertyOrField, INamedTypeSymbol containerClassSymbol,
-            SymbolAnalysisContext context, Func<T, Location> locationSelector, string memberType) where T : ISymbol
-        {
-            var shadowsProperty = GetSelfAndOuterClasses(containerClassSymbol)
+            var selfAndOutterClasses = GetSelfAndOuterClasses(containerClassSymbol);
```

Comment:
From a performance PoV: This is called for each member. Everyone does `GetSelfAndOuterClasses`. 
From a duplication PoV: Everyone also does `.SelectMany(c => c.GetMembers(xxx.Name))` for each member.

Let's move the classes extraction before the `foreach` in `Initialize` (line 53). And the `GetMembers` thing before the switch (line 55)

Reply:
And also `IsStatic` seems to be wanted by everyone, so it can be extracted. It will need extra glitch for IFieldSymbol and IsConst, something like `(...IsStatic || (... is IFieldSymbol {IsConst: true})`

Reply:
I merged all into CheckFieldPropertyEventOrMethod (CheckNamedType is still on its own).